### PR TITLE
Add support for logind idle inhibitors

### PR DESCRIPTION
**Pull request by electrickite**
Adds support for logind idle inhibitor locks. Specifically:

  * swayidle will disable all idle timers when a logind idle inhibitor lock is found

In addition, a few of the logind functions were rearranged and tweaked to support this feature.

This should resolve issue #38 , and would replace/supercede PR #26 and #40

### CODE REVIEW

File: main.c [hunk 1]
```diff
@@ -255,57 +343,76 @@ static int dbus_event(int fd, uint32_t mask, void *data) {
...
-		swayidle_log(LOG_ERROR,
-				"Failed to find session name: %s", error.message);
-		goto cleanup;
+		swayidle_log(LOG_DEBUG,
+				"Failed to find session using PID: %s", error.message);
+		sd_bus_error_free(&error);
+		sd_bus_message_unref(msg);
+
+		const char *env_session = getenv("XDG_SESSION_ID");
+		ret = sd_bus_call_method(bus, DBUS_LOGIND_SERVICE, DBUS_LOGIND_PATH,
```
Comment (melentye) [inline main.c]: Randomly stumbled upon this PR while troubleshooting an issue with IdleHint not working when swayidle is started as a systemd user service.

Have you considered using `auto` magic session name instead of looking up the ID in the environment? That's what [Gnome 3.34 is doing](https://blogs.gnome.org/benzea/2019/10/01/gnome-3-34-is-now-managed-using-systemd/)

That said, this is way outside of my area of expertise so feel free to completely ignore the comment.

Comment (electrickite) [reply main.c]: @melentye Thanks for the suggestion: I like this approach more than pulling from the environment. The latest commit to this branch uses the magic `auto` session. Because this feature is relatively new in systemd, the original lookup by PID is still used as a fallback.

**Comment by electrickite** [comment]
@ddevault  This PR got a little large, but I think it's in a reasonably good place now.

**Comment by ddevault** [comment]
LGTM, adding @emersion for a second review

**Comment by electrickite** [comment]
@emersion Any thoughts on this one?

File: swayidle.1.scd [hunk 2]
```diff
@@ -65,6 +68,19 @@ Sending SIGUSR1 to swayidle will immediately enter idle state.
...
+swayidle responds to the following signals:
+
+*SIGTERM, SIGINT*
+	Clean up timers and exit.
+
+*SIGUSR1*
+	Immediately enter idle state.
+
+*SIGUSR2*
+	Disable/enable idle timers.
```
Comment (gdamjan) [inline swayidle.1.scd]: what does this actually mean? the man page actually doesn't mention "idle timers"

Comment (electrickite) [reply swayidle.1.scd]: It was intended to mean that a SIGUSR2 signal will clear all timeouts. Sending SIGUSR2 again will create them again.

The intent is to allow an easy way to stop swyidle from taking any idle actions, and then re-enable idle actions without having to start a new instance of swayidle.

It could be reworded to something like, "SIGUSR2: Disable all timeout commands. If commands have already been disabled, re-enable them."

Comment (boucman) [reply swayidle.1.scd]: The typical use-case is to disable swayidle from triggering while you're watching a video...

Comment (emersion) [reply swayidle.1.scd]: -1 to this use-case. Use the Sway commands to inhibit idle instead.

Comment (boucman) [reply swayidle.1.scd]: I wonder what convention there is for applications like web-browser to inhibit screensavers...

Comment (emersion) [reply swayidle.1.scd]: There is a [standard Wayland protocol](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/tree/master/unstable/idle-inhibit).

Comment (electrickite) [reply swayidle.1.scd]: @emersion I looked in to using the native sway commands, but they do not cover all use cases. For example, if I want a keyboard shortcut to disable idle, I don't think the sway commands can be used for that. You currently need to kill swayidle and then restart it. 

Adding a signal means you could just do something like: `bindsym $mod+i exec killall -USR2 swayidle` to toggle idle timeouts on or off.

**Comment by electrickite** [comment]
@emersion It looks like this PR needs to be rebased against master. Before I do that, is this still a feature you would consider merging? Are there any blockers?

File: main.c [hunk 3]
```diff
@@ -666,6 +848,7 @@ int main(int argc, char *argv[]) {
 	wl_event_loop_add_signal(state.event_loop, SIGINT, handle_signal, NULL);
 	wl_event_loop_add_signal(state.event_loop, SIGTERM, handle_signal, NULL);
 	wl_event_loop_add_signal(state.event_loop, SIGUSR1, handle_signal, NULL);
+	wl_event_loop_add_signal(state.event_loop, SIGUSR2, handle_signal, NULL);
```
Comment (emersion) [inline main.c]: I'd prefer this to be discussed in a separate pull request. It's not yet clear whether USR2 should toggle, turn off idle timers, or not exist at all.

**Comment by emersion** [comment]
Is there a good reason why this is hidden behind a flag?

**Comment by emersion** [comment]
Otherwise code looks fine.

File: main.c (hunk 3 above)
Comment (electrickite) [reply main.c]: Rebase and subsequent commits have removed SIGUSR2 from this branch.

File: swayidle.1.scd (hunk 2 above)
Comment (electrickite) [reply swayidle.1.scd]: SIGUSR2 has handling has been removed from this PR

**Comment by electrickite** [comment]
@emersion Rebased the branch against master. Also removed SIGUSR2 handling and removed the `-i` option.

File: completions/bash/swayidle [hunk 4]
```diff
@@ -16,6 +16,7 @@ _swayidle()
     -h
     -d
     -w
+    -i
```
Comment (emersion) [inline completions/bash/swayidle]: This should be removed

File: completions/fish/swayidle.fish [hunk 5]
```diff
@@ -10,3 +10,4 @@ complete -c swayidle --condition "__fish_seen_subcommand_from $time_events" --ex
 complete -c swayidle -s h --description 'show help'
 complete -c swayidle -s d --description 'debug'
 complete -c swayidle -s w --description 'wait for command to finish'
+complete -c swayidle -s i --description 'use logind idle inhibitors'
```
Comment (emersion) [inline completions/fish/swayidle.fish]: Ditto

File: completions/zsh/_swayidle [hunk 6]
```diff
@@ -13,6 +13,7 @@ if (($#words <= 2)); then
 			   '(-h --help)'{-h,--help}'[Show help message and quit]' \
 			   '(-d)'-d'[Enable debug output]' \
 			   '(-w)'-w'[Wait for command to finish executing before continuing]'
+			   '(-i)'-i'[Use logind idle inhibitors]'
```
Comment (emersion) [inline completions/zsh/_swayidle]: Ditto

File: main.c [hunk 7]
```diff
@@ -387,6 +506,39 @@ static void register_timeout(struct swayidle_timeout_cmd *cmd,
...
+	state.timeouts_enabled = true;
+	struct swayidle_timeout_cmd *cmd;
+	wl_list_for_each(cmd, &state.timeout_cmds, link) {
+		register_timeout(cmd, cmd->timeout);
+	}
+}
+
+#if HAVE_SYSTEMD || HAVE_ELOGIND
+static void disable_timeouts(void) {
+	if (!state.timeouts_enabled) return;
```
Comment (emersion) [inline main.c]: Style: braces are mandatory

File: main.c [hunk 8]
```diff
@@ -387,6 +506,39 @@ static void register_timeout(struct swayidle_timeout_cmd *cmd,
 	cmd->registered_timeout = timeout;
 }
 
+static void enable_timeouts(void) {
+	if (state.timeouts_enabled) return;
```
Comment (emersion) [inline main.c]: Style: braces are mandatory

**Comment by emersion** [comment]
Can you rebase and squash all of these commits?

File: completions/bash/swayidle (hunk 4 above)
Comment (electrickite) [reply completions/bash/swayidle]: Done

File: main.c (hunk 8 above)
Comment (electrickite) [reply main.c]: Braces added

**Comment by electrickite** [comment]
@emersion Commits squashed and rebased. I believe the latest version addresses your feedback.
